### PR TITLE
(Experiment) Add tags to the schema & export a webfeatures tree

### PR DESCRIPTION
**Pull request by Elchi3**
This draft PR is playing with the idea that compat features have tags. I'm experimenting with a few things here:

## `tags` added to `__compat`

Adding a new optional property at the feature level called "tags". It's always an array. It should always be namespaced for validation purposes. The namespace "type:" is just for illustration of future extensions. The only allowed tag namespace this experiment introduces for now is "webfeatures:".
```json
 "tags": [
   "webfeatures:idle-detection",
   "type:interface"
],
```

## Exporting a new `webfeatures` tree

Adding a new "webfeatures" export structure to the BCD build. It is an array of objects collected from BCD features that are tagged with the "webfeatures:GROUPNAME" tag. The object is a WebFeature object with three properties: id (name of the group), features (array of BCD paths), and compat (a computed compat statement from all of the BCD paths):
```json
"webfeatures": [
  {
   "id": "idle-detection" 
   "features": [
      "api.IdleDetector",
      "api.IdleDetector.IdleDetector",
      "api.IdleDetector.change_event",
      "api.IdleDetector.requestPermission_static",
      "api.IdleDetector.screenState",
      "api.IdleDetector.start",
      "api.IdleDetector.userState",
      "http.headers.Permissions-Policy.idle-detection"
    ],
   "compat": { }
  }
],
```

### View all features tagged at contribution time
To view which features are part of which tags at contribution time, run the traverse script with a new `tag`/`t` argument: `npm run traverse -- -t webfeatures:idle-detection`. This spits out:
```
api.IdleDetector
api.IdleDetector.IdleDetector
api.IdleDetector.change_event
api.IdleDetector.requestPermission_static
api.IdleDetector.screenState
api.IdleDetector.start
api.IdleDetector.userState
http.headers.Permissions-Policy.idle-detection
8
```

### Applying tags to many features at once

Still to be done. Envision a command line tool here.

`npm run tag apply FEATUREPATH`
`npm run tag remove FEATUREPATH` etc.

## Tag validation

Still to be done. Allow-list? Via web-features? Via a file in BCD?

### CODE REVIEW

File: scripts/lib/stringify-and-order-properties.ts [hunk 1]
```diff
@@ -36,7 +36,7 @@ const propOrder = {
       'description',
       'mdn_url',
       'spec_url',
-      'matches',
```
Comment (teoli2003) [inline scripts/lib/stringify-and-order-properties.ts:39]: Good catch: `matches` don't exist (anymore).

Note that `source-file` exists in the schema but is not listed. Should we add it?

**Comment by Elchi3** [comment]
Daniel and I met (twice) about this!

## On the authoring side

We both agreed that instead of naming the whole thing "groups", it would be much better to add a new property "tags" to the schema (alternative we could discuss: "label"). Tags seem more generally useful for future grouping or categorizing of BCD features.

Daniel made a good point above that namespaces will help with introducing tags to BCD. We think that we can introduce tags to BCD now and only allow one namespace "webfeatures" for the moment. Tags without namespaces are not allowed at this point. In the future, we can open up more to potentially other namespaces like "type:", or so. This is to guard us against MDN wiki tags which we had plenty of with no concepts. However, at the same time it will allow us to open the gates some more at a later stage if say we want to group features not just for web-feature purposes but for some other things like MDN landing pages that want to talk about a collection of features for example. 
That said, for the moment, we want to be able add something like this:

```
"tags": [
   "webfeatures:idle-detection"
],
```

- For contributors to view all features that have a certain tag, we will have `npm run traverse -- -t webfeatures:idle-detection`.
- For contributors to apply and remove tags, we want to introduce a command like `npm run tag -- <apply|remove> webfeatures:NAME path.to.feature`
- For contributors and reviewers, we want to provide a detailed documentation what makes a good webfeatures group. This will need close collaboration with how the the web-features project sees groups: https://github.com/web-platform-dx/web-features/tree/main/docs.
- Of course, if there are things in these guidelines which we can automatically validate, then we would add it to a tag linter for "webfeature: tags
- After say 4 weeks of introducing tags to BCD, we want to look at all the webfeatures tags and see if they are okay or if we need even stricter guidelines and linters.

## On the consumer side

As you can see above in this PR, there is also an experiment that plays with a new export tree called "webfeatures". Daniel and I talked about a few things here:
- This new tree is basically a computation from existing BCD trees taking into account constituent features for a group.
- This tree could also be computed externally, BCD doesn't need to provide it by default. The BCD repo could produce a second package or web-features could query into BCD tags and feature and does everything on its own.
- Computing the compat for a group is probably non-trivial. There will be many edge cases (what to do with things like notes, prefixes, etc).
- We need a way to provide overrides if the computed compat data for a group is not satisfactory for some reasons. We believe https://github.com/mdn/browser-compat-data/pull/21119 proposes "override files" but we also think that we could probably come up with a much simpler structure for authors to provide overrides. Also, again, unclear where all this should live. Is it BCD or is it web-features?
- All this said, we need to experiment a lot more on the consumer side and see what feels best.

## Next steps

To move forward, we think it would be great to reach consensus on the first part as soon as possible and introduce tags to BCD so that we can come up with webfeature groups within BCD and quickly get to a lot more webfeature groups, because it will be part of the BCD process and if you add new BCD features you should be adding a new webfeature tag. Also, once we have this in, the usual BCD tree will include tags. While possibly a bit inconvenient for consumers, it would provide everything needed to compute group compat data already.

Then (second priority, can happen in parallel, though), we want to figure out the consumer side and how to best export computed group compat data. The computation of "summerized compat data for a list of features" is one complex part here already. Another part is architectural: where do we do this? In BCD or in web-features or somewhere else?

I will present this topic in the next BCD call. This PR can stay open for the moment, but I will create a new PR that just focuses on the first part: landing the "tags" property in BCD.

Finally, @ddbeck, please let me know if I misrepresented anything here.

**Comment by ddbeck** [comment]
@Elchi3's recap is very good. No objections. I do have one point I want to add/emphasize:

Bottom-up tagging is potentially messy. To move forward on this without digging a hole too deep, we'd start by allowing only tags in the `features`/`webfeatures`/`web-features`/`web-platform-dx/web-features` name space (name TBD), which we'd document as explicitly linking to/from the web-features package's IDs.

This has a few immediate benefits:

- We don't have a _completely_ unconstrained taxonomy. Contributors will be alerted to the fact that we're doing a specific kind of labeling.
- Authors who actually know something about the features in question can start creating new feature groups in a fast, low-friction way (they tag some features while they update the data), without necessarily needing an antecedent on the web-features side first.
- We establish a editorial loop between web-features and BCD. For example, when new tags appear here without a counterpart on web-features, web-features can bless those tags (i.e., I can create a YAML file in web-features to point at the tag here), reject those tags (e.g., create a PR here removing the tag), or begin a deeper investigation (e.g., create a feature on web-features and change what it covers here).
- I'll be forced to write some documentation describing how to create a feature, so there can be guidelines on how to tag them in BCD.

**Comment by ddbeck** [comment]
On a tangential point, I commented in chat that I had expectations about tooling for bulk edits. So it's not lost to time, I'll reproduce that idea here, with commentary:

(When the schema change lands, I'll move this into an issue.)

My dream tool would offer a consistent sort of `<noun> <verb> <args>` interface for bulk modifying BCD entries. This could be a successor to the traverse tool extension proposed above. For tags/labels usage help would look something like this:

```shell
$ dream-tool --help
dream-tool <command> <subcommand> [args]
dream-tool label apply LABELNAME FEATURE […]
dream-tool label remove LABELNAME FEATURE […]
dream-tool label view LABELNAME
```

### Example: apply a tag to many features at once

```shell
$ dream-tool label apply web-features:idle-detection api.IdleDetector api.IdleDetector.IdleDetector …
```

### Example: apply a tag recursively

```shell
$ dream-tool label apply web-features:idle-detection api.IdleDetector 'api.IdleDetector.*'  # quoted due to shell globbing
```

### Example: remove a tag

```shell
$ dream-tool label remove web-features:idle-detection api.setTimeout
```

### Example: view a tag

```shell
$ dream-tool label view web-features:idle-detection
api.IdleDetector
api.IdleDetector.IdleDetector
…
```

**Comment by Elchi3** [comment]
Closing this. Thanks for discussion and the collaborative design work! I'm proposing to introduce tags in https://github.com/mdn/browser-compat-data/pull/21188.

**Pull request closed without merging**